### ISSUE TITLE
Confusing results of custom fine-tuned models

### ISSUE THREAD

**Issue by aloboa** [computational-cell-analytics/micro-sam#837]
This is what I have done:

Custom finetuned Models (with the GUI):
* fine_01.pt : with Segmentation Decoder
* fine_02.pt : without Segmentation Decoder
* best.pt: with Segmentation Decoder

```
micro_sam.automatic_segmentation 
-i imain.tif 
-o imaout1.tif 
-e embeddings_DI17_3_AHE_fine01 
-c fine_01.pt 
-m vit_b 
-d 'cuda' 
--tile_shape 1024 1024 
--halo 256 256 
--pred_iou_thresh 0.7 
--stability_score_thresh 0.9 
--box_nms_thresh 0.1 
--mode amg
--ndim 2 
```
```

micro_sam.automatic_segmentation 
-i imain.tif 
-o imaout2.tif 
-e embeddings_DI17_3_AHE_fine02 
-c fine_02.pt 
-m vit_b 
-d 'cuda' 
--tile_shape 1024 1024 
--halo 256 256 
--pred_iou_thresh 0.7 
--stability_score_thresh 0.9 
--box_nms_thresh 0.1 
--mode amg 
--ndim 2 
```

```
micro_sam.automatic_segmentation 
-i imain.tif  
-o imaout3.tif 
-e embeddings_DI17_2_AHE_256_fine03 
-c best.pt 
-m vit_b 
-d 'cuda' 
--tile_shape 1024 1024 
--halo 256 256 
--mode ais 
--ndim 2 
```

According to your explanations:

> * with fine_01b.pt: You only have the SAM backbone (i.e. the image encoder, prompt encoder and mask decoder). This part of exporting is done to ensure our checkpoints' cross-tool compatibility for SAM!
> * with sam_model/best.pt: You have all finetuned components in here, including the parameters for the additional instance segmentation decoder.
> 

If both *.pt models are only SAM backbones, they should not be affected by the "With segmentation decoder" option. Therefore, fine_01.pt and fine_02.pt should be the same, and thus results imaout1.tif and imaout2.tif should also be the same. But, actually, they are very different.

**Comment by aloboa** [comment]
I understand that we should expect  fine_01.pt and fine_02.pt, and their respective results imaout1.tif and imaout2.tif, to be not identical. However, the problem is that I think that they are too different:

Autosegmentation using a SAM-backbone model created "with segmentation decoder":

![Image](https://github.com/user-attachments/assets/bee1ad6f-98c3-4620-9fc8-4d6c0e9ccada)


Autosegmentation using a SAM-backbone model created "without segmentation decoder":

![Image](https://github.com/user-attachments/assets/758cc203-6c82-4748-98df-be97f8ec70d0)

Maybe you could run a test on your own, with your own data? Simply fine-tune with and without segmentation decoder and autosegment with the respective *.pt models.  Just to make sure the problem is not on my side.

**Comment by anwai98** [comment]
> However, the problem is that I think that they are too different:

Interesting. The thing is, we haven't experimented much with different hyperparameters for AMG-based segmentation (I am assuming the screenshots above are from AMG since that's the one we have talked about the most).

Out of curiousity, which of the above segmentation quality matches closest to your expectations? (the one in blue / red / neither)

> Maybe you could run a test on your own, with your own data?

The data we work with (biological images), it would be hard to observe these changes as AMG did not really improve the segmentation quality with further finetuning (it worked fine on sparsely distributed objects, but that's less likely to assume this in microscopy images).

>  Just to make sure the problem is not on my side.

I still don't think there is any issue from your side. The heuristic with which AMG is supported (by putting a grid of points and then estimating valid segmentations using post-processing and NMS) is quite sensitive. In addition, we introduce another stochastic factor here: two different finetuned models.

I can suggest the following: finetune two models with the exact same training setup (for simplicity, let's say without the additional instance segmentation decoder), export the two different paths with two separate names and then validate your observation. If the difference in segmentations still exists, you can make this a part of your study (as this is often encountered in training / finetuning deep learning models)

PS. @constantinpape Let me know if I am missing something obvious here!

**Comment by aloboa** [comment]
> Out of curiousity, which of the above segmentation quality matches closest to your expectations? (the one in blue / red / neither)

The blue one (without) is a bit better, but neither is really good. The one from a `best.pt `model "with segmentation decoder" is clearly better, although a non-tuned `vit_h` model is the most useful. The reason is that while the fine-tuned one has higher accuracy, it segments mostly the training objects only. While the` vit_h` incurs some on error in the training set, it segments a lot of additional objects (ie, it provides a more general result, which is more useful). I am writing all this now and will let you know.

> I can suggest the following: finetune two models with the exact same training setup (for simplicity, let's say without the additional instance segmentation decoder), export the two different paths with two separate names and then validate your observation. If the difference in segmentations still exists, you can make this a part of your study (as this is often encountered in training / finetuning deep learning models)

Ok, we'll do it.

**Comment by anwai98** [comment]
Thanks for the context. I think I understand the details a bit better. Maybe this will all piece in together if you could share with me some ground-truth for this data? (which you use to finetune your custom models?) (if that's okay with you)

Seeing some image and corresponding ground-truth would help us understand the details a bit better!

**Comment by aloboa** [comment]
> Seeing some image and corresponding ground-truth would help us understand the details a bit better!

[AI-driven processing of SEM images of thin sections of volcanic rocks: a test with Segment Anything for Microscopy](https://diapiro.geo3bcn.csic.es/alobo/blog/microsam_test_SEMGEO_nb.html)

In section "2.Input information"  you can see the 3 original, enhanced and labeled images, just by selecting the tabs.

But I can send you the actual data if you need them.

**Comment by anwai98** [comment]
> In section "2.Input information" you can see the 3 original, enhanced and labeled images, just by selecting the tabs.

I understand it better now. Thanks for sharing the report!

**Comment by anwai98** [comment]
Hi @aloboa,

We made a new release, where we implemented a new finetuning CLI (`micro_sam.train`), which allows you to have a lot more flexibility compared to the GUI.

As our next milestone, we will work on improving the GUI to have some more flexibility. For now (to avoid the issues you faced above), I suggest using python scripts / the finetuning CLI for this.

Let us know if you come across any issues!

**Comment by anwai98** [comment]
Hi @aloboa,

Happy to hear back from you!

If I am not wrong, you already had the installation from source, right? (if yes, just pulling the latest state on the `master` branch should do the trick, otherwise I suggest installing from source for now, in case we need any debugging)
